ARROW-12861: [C++][Compute] Add sign function kernels (#10395)
apache/arrow: docs/source/cpp/compute.rst

+--------------------------+------------+--------------------+---------------------+-------+
| power_checked            | Binary     | Numeric            | Numeric             |       |
+--------------------------+------------+--------------------+---------------------+-------+
| sign                     | Unary      | Numeric            | Int8                | \(1)  |
+--------------------------+------------+--------------------+---------------------+-------+
| sign_with_signed_zero    | Unary      | Numeric            | Int8                | \(2)  |
+--------------------------+------------+--------------------+---------------------+-------+
| subtract                 | Binary     | Numeric            | Numeric             |       |
+--------------------------+------------+--------------------+---------------------+-------+
| subtract_checked         | Binary     | Numeric            | Numeric             |       |
+--------------------------+------------+--------------------+---------------------+-------+

* \(1) Output is an integral representing the signedness of the input.
  For nonzero inputs, output is any of (-1,1) and 0 for zero input.

> * \(2) Similar to ``sign`` but zero input is considered as a signed number.

Reviewer: ```suggestion
* \(2) Similar to ``sign`` but zero is considered signed for floating point inputs.
```

Author: The `sign_with_signed_zero` variant treats 0 as a positive value for integral types, but either -0 or +0 for floating-point types.

Author: Hmm...I agree with this change. The `sign_with_signed_zero` variant only applies to floating-point numbers resulting in either (-1,1). For integral values, the result will be either of (-1,0,1).

+--------------------------+------------+--------------------+---------------------+
| power_checked            | Binary     | Numeric            | Numeric             |
+--------------------------+------------+--------------------+---------------------+

> | sign                     | Unary      | Numeric            | Int8                |

Reviewer 2: Should add a note explaining how the output is computed. You can look at other sections for examples.

Author: I added a "Notes" column to the table, and included notes describing the output values and special case of 0.
